Deck management program, generic report datums, crew record refactor

afterthought2 opened the pull request:
:cl:
rscadd: Adds a deck management program, which gives some information on shuttle movements.
rscadd: Adds flight plans, which can be filed via this program by those with shuttle flight access.
rscadd: If the flight plan includes a roster, the members on the roster can be summoned via automatic comms announcements or batch-emailed.
rscadd: Adds other reports that supply can file about shuttle missions after the mission has departed/returned.
rscadd: Adds a stripped-down report viewer program, currently mostly useful for viewing emailed reports.
/:cl:

It is the year 2562, and NT is commencing a previously unheard-of initiative: digitizing paperwork.

Also some backend changes, notably a refactor of crew records. Reports are now easy to generalize, although the stand-alone program has few features so far.

Will likely make further code improvements, but opening this up for feedback and diff.

To clarify: the only shuttles being logged are Charon, Aquila, Petrov, and Guppy. No logging the Skipjack. None of these reports are necessary to mechanically operate the shuttles or anything like that. You need hangar access to view stuff in the program, supply access to file some reports, and shuttle helm access to do stuff with missions/flight plans.

afterthought2 committed:
deck management, shuttle logs

now with crew records refactor

count text2path

addressed some comments

the getter-setter strikes back

qdel

afterthought2 commented:
Whatever other sins this PR may be committing, touching maps is not one of them. It does touch some torch-only files.

PsiOmegaDelta commented:
Have you ensured this doesn't break the Fake Crew Records one can acquire through the Traitor uplink, which is why we reverted the last 2 attempts of redoing crew records?

PsiOmegaDelta commented:
Currently the bot makes the same changes for maps as with sprites and sounds, i.e. if you've made changes in the ´maps´ subfolder.

PsiOmegaDelta commented on code/modules/shuttles/shuttle_log.dm:
```diff
@@ -0,0 +1,111 @@
+
+/datum/shuttle_log/New(datum/shuttle/given_shuttle)
+	..()
+	if(istype(given_shuttle) && given_shuttle.logging_home_tag)
```
Should it be possible for a shuttle log to not receive a proper shuttle datum?

afterthought2 replied:
Will drop check.

PsiOmegaDelta commented on code/game/jobs/access_datum.dm:
```diff
@@ -297,7 +297,12 @@
 
-// /var/const/free_access_id = 51
+/var/const/access_hangar = 51
+/datum/access/hangar
```
You couldn't figure out how to make an access checks system that doesn't rely on a hardcoded value?

afterthought2 replied:
I could probably make it a map variable, if that's preferable. Not used to map-only code, so just tell me what would be best.

PsiOmegaDelta commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -658,3 +658,11 @@ proc/is_blind(A)
+		if(mob.real_name == real_name)
+			if(!mob.mind)
+				return
+			return mob.mind.initial_email_login["login"]
```
..e-mails are stored in the mob mind?
So if an admin, wizard spell, or otherwise switch two mobs the mobs will also have new e-mail addresses?
Who thought this was a good idea?

afterthought2 replied:
There are alternative methods of searching for emails, like searching the list of all email logins for one which might match someone's name. This actually seems to be the most reliable.

Maybe to clarify, there is a global list of all email logins, and it's possible, given a name, to guess the login and search the list. This is unreliable because if someone's login is misguessed (say because the algorithm is changed, or a way of changing logins is implemented), it fails. On the other hand, it's probably more satisfactory from a philosophical point of view.

It may be time to have email datums store mobs' original names, but that seemed outside the scope of the PR (you'd have to think through if/when to update the datums).

PsiOmegaDelta commented on code/modules/shuttles/shuttle_log.dm:
```diff
@@ -0,0 +1,111 @@
+	var/list/datum/shuttle_mission/missions = list()  //The actual log entries.
+	var/datum/shuttle_mission/current_mission         //The current mission, planned or ongoing.
+	var/home_base                                     //The landmark tag from which missions originate.
+	var/ID_ticker = 1                                 //Used to assign unique IDs to missions.
```
Use `sequential_id(key)`

afterthought2 replied:
Will look into that for ID stuff. Thanks.

afterthought2 commented:
I tested fake crew records, and they seem to work as intended (comms announcement, fake duplicate record with some stuff copied over). I did change/simplify the code for them, but from what I can understand they should be functionally identical. If there is something more subtle I'm missing, I'm happy to look into it.

comma commented:
That's a lot of noise in that diff.

I think the thing you done to crew records is very gross and is unnecessary. Why turn crew records around if you just need the record datums? Now we got all fields stored in a list AND in bunch of vars. Does clone() even work now? It copies all the fields, but does it set those vars? State is sinful because it causes issues like that where you have to remember to update all the places. 

Not a fan of mass rename of previously existing thing to accomodate newly added thing. Neither 'record_field' nor 'report_field' properly cover either use case so why change.

As I was saying in discord, you seem to just need a subtype of record field that has an ID.
Only logic you needed from crew records were clone and getters setters (which you didn't have use anyway since they were getting by type). Just add get_by_name/id and now you don't need to touch crew records at all.

CakeQ commented:
This sounds amazing. Any preview images?

afterthought2 commented:
@comma, can't really argue with you, but I also couldn't figure out how to make it work without either some major changes to crew records or a lot of duplicate code.

I could easily get rid of the vars and return the getter setter, which would reduce the number of files in the diff. I agree the vars are pretty bad, that's mostly what I was talking about when I said I still wanted to make code improvements. I could even index by type there if necessary, though the macro stuff will still need to be doing add_field to avoid subtypesof. The vars are just there to get a version running. (for what it's worth, cloning didn't work in the original code either, but that's a valid point too. EDIT: cloning should work correctly).

But making crew records a subtype of report lets them share code, like generate_nano_data whenever you shove them into nanoui, or the access control stuff.

Anyway, I'll think some more on how to do this better, and either make some adjustments or close it.

comma commented:
I might try to hack something myself, but dunno if will have energy.
On positive side i thibk the feature itself is nice

afterthought2 commented:
Goodbye loads of vars, welcome back preprocessor. The comment said it was my friend.

Crew records are back to only storing the fields in a list and fetching them with getters/setters that internally search by type. The details are changed slightly to make subtyping fields elsewhere not break stuff accidentally, but same in spirit. Getters/setters have the same names as before.

Still lots of files in the diff because I changed the type path of crew records, but I do argue that's a positive change that allows for better code sharing. Also some changes elsewhere that simplify I/O stuff with shared procs, fix random bugs/strangeness, and kill text2path shenanigans have been kept.

The actual length of the diff is partly because I moved a file.

comma commented on code/modules/shuttles/shuttle_log.dm:
```diff
@@ -0,0 +1,109 @@
+	if(istype(report, /datum/computer_file/report/recipient/shuttle))
+		var/old_report = locate(report.type) in mission.other_reports
+		if(old_report)
+			mission.other_reports -= old_report
```
does anything else reference old_report? might cause them to pile up in nullspace undeleted if so. IMO just qdel it for good measure if you're not using it afterwards.

afterthought2 replied:
Made it qdel. It shouldn't have any references that aren't temporary (would be cleared after nanoui update), but better be safe.

comma commented on code/modules/shuttles/shuttle_log.dm:
```diff
@@ -0,0 +1,109 @@
+/datum/shuttle_log/proc/shuttle_returned()
+	if(!current_mission || (current_mission.stage != SHUTTLE_MISSION_STARTED))
+		current_mission = null
+		CRASH("Shuttle returned, but mission stage was incorrect or no mission was logged.")
```
Is it breaking stuff if you don't log a mission? Would be silly if like, an antag stealing a shuttle would generate a runtime (cause that's what CRASH does)

afterthought2 replied:
It is. This should never happen, and if it does, it's a code oversight on my part regarding shuttle movements. I want it to runtime so that I can fix the problem.

It should autocreate a mission (if one isn't set up) whenever a shuttle leaves, and update it when it comes back.

afterthought2 commented:
Squashed.
